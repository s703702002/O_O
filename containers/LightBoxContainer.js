import React from 'react';
import { connect } from 'react-redux';
import LightBox from '../components/LightBox';
import { removeLightBoxMessage } from '../action';

const LightBoxWithConnect = ({ lightBoxMessage, removeLightBoxMessage: removeLightBox }) => {
  if (lightBoxMessage.length > 0) {
    return <LightBox message={lightBoxMessage} removeLightBox={removeLightBox} />;
  }
  return null;
};

const mapStateToProps = (state) => {
  const { lightBoxMessage } = state;
  return {
    lightBoxMessage,
  };
};

export default connect(
  mapStateToProps,
  { removeLightBoxMessage },
)(LightBoxWithConnect);
