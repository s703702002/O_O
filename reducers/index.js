import { combineReducers } from 'redux';

const login = (state = { status: 'init' }, action) => {
  switch (action.type) {
    case 'LOGIN_REQUEST':
      return {
        status: 'loading',
        username: null,
        message: null,
      };
    case 'LOGIN_SUCCESS':
      return {
        status: 'logined',
        username: action.response.username,
        message: null,
      };
    case 'LOGIN_ERROR':
      return {
        status: 'loginerr',
        username: null,
        message: action.error,
      };
    case 'LOG_OUT':
      return {
        status: 'init',
        username: null,
        message: null,
      };
    default:
      return state;
  }
};

const loginBoxOpen = (state = false, action) => {
  switch (action.type) {
    case 'OPEN_LOGIN_BOX':
      return true;
    case 'CLOSE_LOGIN_BOX':
      return false;
    default:
      return state;
  }
}

const AppStore = combineReducers({
  login,
  loginBoxOpen,
});

export { login, loginBoxOpen };
export default AppStore;
