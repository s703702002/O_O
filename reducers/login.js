import { clone } from '../utilis';

const login = (state = { status: 'init' }, action) => {
  const newShoppings = !state.shoppings ? null : clone(state.shoppings);
  switch (action.type) {
    case 'LOGIN_REQUEST':
      return {
        status: 'loading',
        username: null,
        message: null,
        shoppings: [],
      };
    case 'LOGIN_SUCCESS':
      return {
        status: 'logined',
        username: action.response.name,
        message: null,
        shoppings: action.response.shoppings,
      };
    case 'LOGIN_ERROR':
      return {
        status: 'loginerr',
        username: null,
        message: action.error,
        shoppings: [],
      };
    case 'LOGIN_CANCEL':
      return {
        state: 'init',
        username: null,
        message: null,
        shoppings: [],
      };
    case 'LOG_OUT':
      return {
        status: 'init',
        username: null,
        message: null,
        shoppings: [],
      };
    case 'ADD_TO_CART':
      newShoppings.push(action.product);
      return {
        ...state,
        shoppings: newShoppings,
      };
    default:
      return state;
  }
};

export default login;
