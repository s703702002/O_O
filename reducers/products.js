const products = (
  state = {
    message: null,
    products: [],
  },
  action,
) => {
  switch (action.type) {
    case 'RECEIVE_PRODUCTS':
      return {
        products: action.response,
        message: null,
      };
    case 'GET_PRODUCTS_ERROR':
      return {
        products: [],
        message: action.error,
      };
    default:
      return state;
  }
};

export default products;
