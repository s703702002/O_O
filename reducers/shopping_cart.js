const isNewItem = (state, productId) =>
  !state.some(item => item.product.id === productId);

const handleAddToCart = (state, action) => {
  const { product, count } = action;
  const productId = product.id;
  // 若是不在購物車的產品 就新增一筆資料
  if (isNewItem(state, productId)) {
    return [...state, {
      product,
      count,
    }];
  }
  return state.map((item) => {
    const order = { ...item };
    if (order.product.id === productId) order.count += count;
    return order;
  });
};

const handleReduceCartItem = (state, action) => {
  const { productId } = action;

  return state.map((item) => {
    const order = { ...item };
    if (order.product.id === productId) order.count -= 1;
    return order;
  });
};

const shoppingsCarts = (state = [], action) => {
  switch (action.type) {
    case 'LOGIN_SUCCESS':
      return [...action.response.shoppings];
    case 'LOG_OUT':
      return [];
    case 'ADD_TO_CART':
      return handleAddToCart(state, action);
    case 'REMOVE_SHOPPING_CART_ITEM':
      return state.filter(item => item.product.id !== action.productId);
    case 'REDUCE_CART_ITEM':
      return handleReduceCartItem(state, action);
    default:
      return state;
  }
};

export default shoppingsCarts;
