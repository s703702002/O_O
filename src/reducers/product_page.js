const productPage = (
  state = {
    message: null,
    product: {},
  },
  action,
) => {
  switch (action.type) {
    case 'RECEIVE_PRODUCT':
      return {
        product: action.response,
        message: null,
      };
    case 'GET_PRODUCT_ERROR':
      return {
        product: {},
        message: action.error,
      };
    default:
      return state;
  }
};

export default productPage;
