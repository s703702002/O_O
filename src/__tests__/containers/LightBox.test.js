import React from 'react'
import { createStore } from 'redux'
import { render, cleanup } from 'react-testing-library'
import { Provider } from 'react-redux'
import { lightBoxMessage } from '../../reducers'
import LightBoxContainer from '../../containers/LightBoxContainer'

afterEach(cleanup);

const mockInitState = {
  lightBoxMessage: "hello world"
};

function renderWithRedux(
  ui,
  { initialState, store = createStore(lightBoxMessage, initialState) } = {}
){
  return {
    ...render(
      <Provider store={store}>
        {ui}
      </Provider>,
      store
    )
  }
};

test('<LightBoxContainer /> render', () => {
  const { asFragment } = renderWithRedux(<LightBoxContainer/>, {
    initialState: mockInitState
  })

  expect(asFragment()).toMatchSnapshot()
});
