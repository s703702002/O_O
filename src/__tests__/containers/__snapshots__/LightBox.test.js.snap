// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LightBoxContainer /> render 1`] = `
<DocumentFragment>
  <div
    class="modal light_box"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered"
      role="document"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-body"
        >
          <i
            class="material-icons text-warning"
          >
            error_outline
          </i>
          <p
            class="text-warning font-weight-bold message"
          >
            hello world
          </p>
          <button
            class="btn btn-outline-info"
          >
            知道了
          </button>
        </div>
      </div>
    </div>
    <div
      class="mask"
      role="document"
    />
  </div>
</DocumentFragment>
`;
